eth/db/chain_gaps.py

```diff
@@ -12,16 +13,74 @@ class GapChange(enum.Enum):
     NewGap = enum.auto()
     GapFill = enum.auto()
     GapSplit = enum.auto()
-    GapShrink = enum.auto()
+    GapLeftShrink = enum.auto()
+    GapRightShrink = enum.auto()
     TailWrite = enum.auto()
 
 
-GAP_WRITES = (GapChange.GapFill, GapChange.GapSplit, GapChange.GapShrink)
+GAP_WRITES = (
+    GapChange.GapFill,
+    GapChange.GapSplit,
+    GapChange.GapLeftShrink,
+    GapChange.GapRightShrink,
+)
 GENESIS_CHAIN_GAPS = ((), BlockNumber(1))
 
 GapInfo = Tuple[GapChange, ChainGaps]
 
 
+@to_tuple
+def _join_overlapping_gaps(unjoined_gaps: Tuple[BlockRange, ...]) -> Iterable[BlockRange]:
+    """Input must already be sorted"""
+    unyielded_low = None
+    unyielded_high = None
+    for low, high in unjoined_gaps:
+        if unyielded_high is not None:
+            if low < unyielded_low:
+                raise ValidationError(f"Unsorted input! {unjoined_gaps!r}")
+            elif unyielded_low <= low <= unyielded_high + 1:
+                unyielded_high = max(high, unyielded_high)
+                continue
+            else:
+                yield unyielded_low, unyielded_high
+
+        unyielded_low = low
+        unyielded_high = high
+
+    if unyielded_high is not None:
+        yield unyielded_low, unyielded_high
+
+
+def insert_gap(decanonicalized: BlockNumber, base_gaps: ChainGaps) -> ChainGaps:
+    current_gaps, tip_child = base_gaps
+
+    if tip_child <= decanonicalized:
+        return base_gaps
+
+    new_raw_gaps = current_gaps + ((decanonicalized, decanonicalized), )
+
+    # join overlapping gaps
+    joined_gaps = _join_overlapping_gaps(sorted(new_raw_gaps))
+
+    # is the last gap overlapping with the tip child? if so, merge it
+    if joined_gaps[-1][1] + 1 >= tip_child:
+        return joined_gaps[:-1], joined_gaps[-1][0]
+    else:
+        return joined_gaps, tip_child
+
+
+def is_block_number_in_gap(block_number: BlockNumber, gaps: ChainGaps) -> bool:
+    gap_ranges, tip_child = gaps
+    for low, high in gap_ranges:
+        if low > block_number:
+            return False
+        elif high >= block_number:
+            return True
+        # this range was below the block number, continue looking at the next range
+
+    return block_number >= tip_child
+
+
 def calculate_gaps(newly_persisted: BlockNumber, base_gaps: ChainGaps) -> GapInfo:
```

Comment:
probably need to rename this to distinguish more clearly from `insert_gap`. Also noticed a lot of contortion in the naming. Maybe this is: `gap_after_new_header` and `insert_gap` -> `gap_after_dropping_header`, or something like that.

Otherwise I found myself noting everywhere: insert header == drop gap, and drop header == insert gap, to try to avoid confusion.

Reply:
I guess, I'd be fine with either `gap*s*_after_new_header` or `gapinfo_after_new_header`